### ISSUE TITLE
interface-mtu dhcp option causes network initialization loop

### ISSUE THREAD

**Issue by travisgroth** [coreos/bugs#1827]
# Issue Report #

### Container Linux Version ###

```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1313.0.0
VERSION_ID=1313.0.0
BUILD_ID=2017-02-03-0826
PRETTY_NAME="Container Linux by CoreOS 1313.0.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```

### Environment ###

PXE booted Dell R620, Bare Metal

### Expected Behavior ###

Upon system boot DHCP initializes the network and adjusts interface MTU to interface-mtu (or ignores it), then pulls down cloud-config and ignition configs.

### Actual Behavior ###

System loops network initialization until switch shuts the port for excessive flapping.  The loop appears to be related to changing the interface MTU from 1500->9000 and then back for some reason.  This is difficult to work around on the PXE boot version since you don't have an opportunity to override the network settings until the network comes up to fetch a new link/service/unit file.

Here's the output we get:

![image](https://cloud.githubusercontent.com/assets/6198562/23286503/fa18215c-fa05-11e6-93eb-1d43a66d7702.png)

```
[   81.443118] pps pps0: new PPS source ptp2
[   81.493624] ixgbe 0000:01:00.1: registered PHC device on eno2
[   81.699834] IPv6: ADDRCONF(NETDEV_UP): eno2: link is not ready
[   81.837972] pps pps1: new PPS source ptp3
[   81.886252] ixgbe 0000:01:00.0: registered PHC device on eno1
[   81.955520] random: crng init done
[   82.076654] IPv6: ADDRCONF(NETDEV_UP): eno1: link is not ready
[   82.262528] IPv6: ADDRCONF(NETDEV_UP): eno4: link is not ready
[   82.437612] IPv6: ADDRCONF(NETDEV_UP): eno3: link is not ready
[   89.462012] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[   89.549600] IPv6: ADDRCONF(NETDEV_CHANGE): eno1: link becomes ready
[   91.867136] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[   92.444539] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  111.805963] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  114.331919] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  114.907424] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  122.610745] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  124.848328] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  125.426898] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  137.582287] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  140.687268] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  141.266101] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  148.932146] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  150.249229] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  150.836726] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  158.486254] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  161.589224] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  162.169018] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  175.220496] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  177.690994] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  178.272223] ixgbe 0000:01:00.0 eno1: changing MTU from 9000 to 1500
[  197.096359] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
[  197.939198] ixgbe 0000:01:00.0 eno1: changing MTU from 1500 to 9000
[  205.560611] ixgbe 0000:01:00.0 eno1: NIC Link is Up 10 Gbps, Flow Control: None
```

In this exact scenario, we are using intel ixgbe with 10G copper NICs.  The 10gb link negotiation time on these can vary to up to 10 seconds in our experience.  Our RHEL systems have 'LINKDELAY=10' set in their network configuration to prevent DHCP from timing out.

### Reproduction Steps ###

  1. Have 10G copper network with jumbo frames enabled
  2. Boot PXE CoreOS image on network with DHCP configured to hand out non-default interface-mtu option (eg, 9000)
  3. Observe network init loop as pictured above

### Other Information ###

If I jump onto the console and explicitly run:

```
ip link set mtu 9000 dev eno1
```

It ceases the loop and allows the network to fully initialize.  It seems like the change from default to jumbo forces the link to reset and it doesn't come back fast enough, but I'm speculating on the internals a bit.  

I also tried setting netpoll.carrier_timeout=30 at pxe boot time but that doesn't appear to help.  I can't find any other options that might impact this from the kernel command line.

Any recommendations on how I can work around this or address it?  I can provide any additional testing or data as needed.

**Comment by travisgroth** [comment]
More info.

1. This does not happen in a virtualized environment, which is why I believe the link negotiation delay is the culprit. 

2. Did a bit more digging into bootengine.  Setting:
```
ip=:dhcp:9000
```
as a kernel command line option avoids this scenario if I don't have any secondary network unit file.

It *still* avoids the scenario if I have a unit file (set up by ignition) that honors the dhcp mtu:
```
$ cat /etc/systemd/network/00-e.network
[Match]
Name=eno1

[Network]
DHCP=yes

[DHCP]
UseMTU=true
Hostname=coreos0
ClientIdentifier=mac
```

The scenario is reproduced again - but only at the custom network unit file stage -  if I set 
```
ip=:dhcp:1500
```

**Comment by travisgroth** [comment]
Bump.  This has been open since February as a P1.  Is there more info I can provide?

**Comment by bgilbert** [comment]
Thanks for the detailed report.  I'm looking into this issue and will get back to you.

**Comment by bgilbert** [comment]
`netpoll.carrier_timeout` is for netconsole, and won't help here.

Here's what seems to be happening:

1. systemd-networkd obtains a lease and sets the MTU.
2. This causes some network drivers, including e1000 and ixgbe but not virtio, to reset the NIC.
3. systemd-networkd receives a netlink message that the NIC has gone down, discards the lease, and changes the MTU back to its previous value.
4. The NIC comes back up and systemd-networkd requests a lease.
5. goto 1

`ip=:dhcp:9000` avoids the problem because systemd-networkd finds the MTU is already 9000 and so doesn't change it.

## Reproduction steps

As root (assuming a Fedora system):

```sh
brctl addbr testbr
echo "allow testbr" >> /etc/qemu/bridge.conf
firewall-cmd --change-interface=testbr --zone=trusted
ip link set testbr up
ip addr add 192.168.100.1/24 dev testbr
dnsmasq -a 192.168.100.1 -F 192.168.100.2,192.168.100.100 -O 26,1600 -d
```

As non-root:

```sh
./coreos_production_qemu.sh -net bridge,vlan=1,br=testbr -net nic,vlan=1,model=e1000
```

**Comment by ajeddeloh** [comment]
I've filed a bug upstream https://github.com/systemd/systemd/pull/6594

**Comment by travisgroth** [comment]
@ajeddeloh is there any workaround for this?  I'm actually now completely blocked by this as it looks like ip=:dhcp:9000 doesn't appear to work properly anymore.  The system boots and is pingable, but it never actually fetches the ignition config - there's no hit in the matchbox server log at all.  Since the initial report, I've moved on to version 1465.6.0.

**Comment by bgilbert** [comment]
`ip=:dhcp:9000` _is_ the workaround.  It shouldn't have stopped working in 1465, though.

What is your kernel command line?  Does the machine boot successfully or fail to an emergency shell?

**Comment by travisgroth** [comment]
@bgilbert very odd.  I just rolled back to 1353.6.0 and yes I agree this doesn't look like it worked as of then either.  I have a feeling I hadn't been actually trying to do anything with ignition yet.  I may have only been working with cloud config.  

The system hangs for some period and eventually reboots.

Here's where I get stuck with ip=:dhcp:9000 (or ip=eth2:dhcp:9000 as on the physical system I found this on):

![image](https://user-images.githubusercontent.com/6198562/30671300-377c4816-9e34-11e7-8295-1759771c1b30.png)

Eventually it moves past that:
![image](https://user-images.githubusercontent.com/6198562/30671342-69cf39cc-9e34-11e7-958f-81e6bbd6bc2d.png)

Here's the corresponding matchbox rendered ipxe config:
````
#!ipxe
kernel /assets/coreos/1353.6.0/coreos_production_pxe.vmlinuz ip=:dhcp:9000 coreos.config.url=http://matchbox01/ignition?type=ignition-config&key=master-diskprep coreos.first_boot=yes console=tty0 console=ttyS0 coreos.oem=pxe coreos.autologin=tty1 coreos.autologin=ttyS0 coreos.autologin
initrd /assets/coreos/1353.6.0/coreos_production_pxe_image.cpio.gz
boot
````


*if* I remove the coreos.config.url parameter, it boots (though, obviously, with no ignition run).  Here's the rendered ipxe:

````
#!ipxe
kernel /assets/coreos/1353.6.0/coreos_production_pxe.vmlinuz ip=:dhcp:9000 coreos.first_boot=yes console=tty0 console=ttyS0 coreos.oem=pxe coreos.autologin=tty1 coreos.autologin=ttyS0 coreos.autologin
initrd /assets/coreos/1353.6.0/coreos_production_pxe_image.cpio.gz
boot
````

I tried adding systemd.journald.max_level_console=debug but I'm not seeing anything useful.  I can move this into a new issue as it may not be directly related (only exposed by the workaround) now that I'm looking closer.  Let me know.

**Comment by bgilbert** [comment]
The eventual reboot means the system dropped to an emergency shell, which times out after 5 minutes.  This is probably because Ignition failed.  You're not seeing any of that output because it's going to the primary console (`ttyS0`) and you're looking at the VGA console instead.  Try removing `console=ttyS0` from the command line.  Once you're prompted to start an emergency shell, hit Enter and then run `journalctl -t ignition` to see the error logs.

**Comment by travisgroth** [comment]
*sigh* I just realized that the short name in the ignition URL (which worked without ip=xxxxx) might not work if the DNS domain isn't picked up with that config.  I'm triple checking some other changes but I believe that was the issue.

>You're not seeing any of that output because it's going to the primary console (ttyS0) and you're looking at the VGA console instead

I see.  I'll fix that going forward.  Thank you.

**Comment by ajeddeloh** [comment]
Closed via https://github.com/coreos/coreos-overlay/pull/2967

**Comment by bgilbert** [comment]
This should be fixed in 1662.0.0.  Thanks for the report!

**Comment by bgilbert** [comment]
The upstream fix is https://github.com/systemd/systemd/pull/7460.
